### PR TITLE
Add shuttle data dependencies

### PR DESCRIPTION
**Pull request by digitalcora**
**Asana Ticket:** [Shuttles | Establish and wire up API dependencies](https://app.asana.com/0/477545582537986/1149842993086361/f)

I wanted to get this up as early as possible, so there are no tests of the new module yet 😓

It is intended as a one-stop shop for the Shuttles tab specifically. Lists of `shapes` and `stops` are returned, which correspond to polylines and markers we can draw on the map (and `stops` can also be used to build the stop selector for the detail view).

* The stop selector can be built from `stops` of type `rail_affected`
* The map markers can be drawn from `stops` of type `rail_unaffected` and `shuttle`
* The shuttle and rail lines can be drawn from `shapes` according to `is_shuttle_route`

<details>
<summary>Example data structure for the Red Line diversion</summary>
<pre>
%Site.ShuttleService{
   shapes: [
     %Site.ShuttleService.Shape{
       direction_id: 0,
       id: "931_0009-70071-70081-0",
       is_shuttle_route: false,
       polyline: "}nwaG~|eqLGyNIqAAc@S_CAEWu@g@}@u@k@u@Wu@OMGIMISQkAOcAGw@SoDFkCf@sUXcJJuERwHPkENqCJmB^mDn@}D??D[TeANy@\\iAt@qB`AwBl@cAl@m@b@Yn@QrBEtCKxQ_ApMT??R?`m@hD`Np@jAF|@C`B_@hBi@n@s@d@gA`@}@Z_@RMZIl@@fBFlB\\tAP??~@L^?HCLKJWJ_@vC{NDGLQvG}HdCiD`@e@Xc@b@oAjEcPrBeGfAsCvMqVl@sA??jByD`DoGd@cAj@cBJkAHqBNiGXeHVmJr@kR~@q^HsB@U??NgD"
     },
     %Site.ShuttleService.Shape{
       direction_id: 0,
       id: "KendallToBroadway-S",
       is_shuttle_route: true,
       polyline: "}|paGtc{pLMhG_BWeDaCtFaQzAyCt@mF`DgbAv@cFCaAc@cBQ{l@t@eFjAoCdAkAlB}@t@KfBFtD|BxLnM`CrBbI~CrE`ALuFo@qO@kD^gJb@_BbAiIoCwAfGgPlCmIf@Lf@ItEwHbVfQvE|CvDhBhO~QPv@iKD"
     },
     %Site.ShuttleService.Shape{
       direction_id: 0,
       id: "931_0009-70071-70081-2",
       is_shuttle_route: false,
       polyline: "k`maGziupL??bnAm@h@T??xF|BpBp@^PLBXAz@Yl@]l@e@|B}CT[p@iA|A}BZi@zDuF\\c@n@s@VObAw@^Sl@Yj@U\\O|@WdAUxAQRCt@E??xAGrBQZAhAGlAEv@Et@E~@AdAAbCGpCA|BEjCMr@?nBDvANlARdBb@nDbA~@XnBp@\\JRH??|Al@`AZbA^jA^lA\\h@P|@TxAZ|@J~@LN?fBXxHhApDt@b@JXFtAVhALx@FbADtAC`B?z@BHBH@|@f@RN^^T\\h@hANb@HZH`@H^LpADlA@dD@jD@x@@b@Bp@HdAFd@Ll@F^??n@rDBRl@vD^pATp@Rb@b@z@\\l@`@j@p@t@j@h@n@h@n@`@hAh@n@\\t@PzANpAApBGtE}@xBa@??xB_@nOmB`OgBb@IrC[p@MbEmARCV@d@LH?tDyAXM"
     },
     %Site.ShuttleService.Shape{
       direction_id: 0,
       id: "933_0009-70071-70081-0",
       is_shuttle_route: false,
       polyline: "}nwaG~|eqLGyNIqAAc@S_CAEWu@g@}@u@k@u@Wu@OMGIMISQkAOcAGw@SoDFkCf@sUXcJJuERwHPkENqCJmB^mDn@}D??D[TeANy@\\iAt@qB`AwBl@cAl@m@b@Yn@QrBEtCKxQ_ApMT??R?`m@hD`Np@jAF|@C`B_@hBi@n@s@d@gA`@}@Z_@RMZIl@@fBFlB\\tAP??~@L^?HCLKJWJ_@vC{NDGLQvG}HdCiD`@e@Xc@b@oAjEcPrBeGfAsCvMqVl@sA??jByD`DoGd@cAj@cBJkAHqBNiGXeHVmJr@kR~@q^HsB@U??NgD"
     },
     %Site.ShuttleService.Shape{
       direction_id: 0,
       id: "933_0009-70071-70081-2",
       is_shuttle_route: false,
       polyline: "k`maGziupL??bnAm@h@T??xF|BpBp@^PLBXAz@Yl@]l@e@|B}CT[p@iA|A}BZi@jBeDnAiBz@iAf@k@l@g@dAs@fAe@|@WpCe@l@GTCRE\\G??~@O`@ELA|AGf@A\\CjCGrEKz@AdEAxHY|BD~@JjB^fF~AdDbA|InCxCv@zD|@rWfEXDpB`@tANvAHx@AjBIx@M~@S~@a@fAi@HEnA{@fA{@|HuI|DwEbDqDpLkNhCyClEiFhLaN`@c@f@o@RURUbDsDbAiA`AgAv@_AHKHI~E}FdBoBfAgAfD{DxDoE~DcF|BkClAwALODEJOJK|@gATWvAoA`Au@fAs@hAk@n@QpAa@vDeAhA[x@Yh@Wv@a@b@YfAaAjCgCz@aAtByBz@{@??|FaGtCaDbL{LhI{IzHgJdAuAjC{CVYvAwA??JIl@a@NMNM\\[|AuArF_GlPyQrD_ErAwAd@e@nE{ErDuD\\a@nE_FZYPSRUvL{Mv@}@Z[JILKv@m@z@i@fCkAlBmAl@[t@[??h@WxBeAp@]dAi@p@YXIPEXKDALENEbAQl@Gz@ChADtAL~ARnCZbGx@xB`@TDL@PBzAVjIvA^FVDVB|@NjHlAlPnCnCd@vBXhBNv@JtAPL@|BXrAN??`@FRBj@Bp@FbADz@?dAIp@I|@Mx@Q`AWhAYlBs@pDaBzAs@nBgAZQJGJGhAs@RKVMNKTMf@YdHcEzBmApAw@`GmDLI@AHGlEwClAi@hA_@v@Up@ObB]z@Kr@Ir@EZCpA?dCRf@DpAHvANrE`@bDTr@DfMdA`CJvBRn@DnCLnBPfAFV@"
     },
     %Site.ShuttleService.Shape{
       direction_id: 1,
       id: "931_0010-70082-70072-0",
       is_shuttle_route: false,
       polyline: "qsaaGfrvpLYLuDxAI?e@MWASBcElAq@LsCZc@HaOfBoOlByB^??yB`@uE|@qBFqA@{AOu@Qo@]iAi@o@a@o@i@k@i@q@u@a@k@]m@c@{@Sc@Uq@_@qAm@wDCSo@sD??G_@Mm@Ge@IeACq@Ac@Ay@AkDAeDEmAMqAI_@Ia@I[Oc@i@iAU]_@_@SO}@g@IAIC{@CaB?uABcAEy@GiAMuAWYGc@KqDu@yHiAgBYO?_AM}@KyA[}@Ui@QmA]kA_@cA_@aA[}Am@??SI]KoBq@_AYoDcAeBc@mASwAOoBEs@?kCL}BDqC@cCFeA@_A@u@Dw@DmADiAF[@sBPyAF??u@DSByAPeAT}@V]Nk@Tm@X_@RcAv@WNo@r@]b@{DtF[h@}A|Bq@hAUZ}B|Cm@d@m@\\{@XY@MC_@QqBq@{F_C??g@ScnAl@"
     },
     %Site.ShuttleService.Shape{
       direction_id: 1,
       id: "BroadwayToKendall-S",
       is_shuttle_route: true,
       polyline: "}`maGzhupLnD{GmEuFgBsBwDiBwE}CcVgQuEvHg@Hg@WWT_KlXaA~E}@xBoAgCwRpBUiEiCbCoFhAO~AF|ByA~HWd@yBz@uBfB}AxDm@jEXvm@o@vCLbGgDjdA[rB_AfBz@tBi@xK"
     },
     %Site.ShuttleService.Shape{
       direction_id: 1,
       id: "931_0010-70082-70072-2",
       is_shuttle_route: false,
       polyline: "}{paGtyzpLOfD??ATIrB_Ap^s@jRWlJYdHOhGIpBKjAk@bBe@bAaDnGkBzD??m@pAwMpVgArCsBdGkEbPc@nAYb@a@d@eChDwG|HMPEFwCzNK^KVMJIB_@?iAO??kAOmB]gBGm@A[HSL[^a@|@e@fAo@r@iBh@aB^}@BkAGaNq@am@iDQ???sMUyQ~@uCJsBDo@Pc@Xm@l@m@bAaAvBu@pB]hAOx@UdAKr@??i@dD_@lDKlBOpCQjESvHKtEYbJg@rUGjCRnDFv@NbAPjAHRHLLFt@Nt@Vt@j@f@|@Vt@@DR~B@b@HpAFxN"
     },
     %Site.ShuttleService.Shape{
       direction_id: 1,
       id: "933_0010-70082-70072-0",
       is_shuttle_route: false,
       polyline: "iyr`GzmjpLWAgAGoBQoCMo@EwBSaCKgMeAs@EcDUsEa@wAOqAIg@EeCSqA?[Bs@Ds@H{@JcB\\q@Nw@TiA^mAh@mEvCIFA@MHaGlDqAv@{BlAeHbEg@XULOJWLSJiAr@KFKF[PoBfA{Ar@qD`BmBr@iAXaAVy@P}@Lq@HeAH{@?cAEq@Gk@CSCc@G??qAO}BYMAuAQw@KiBOwBYoCe@mPoCkHmA}@OWCWE_@GkIwA{AWQCMAUEyBa@cGy@oC[_BSuAMiAE{@Bm@FcAPODMDE@YJQDYHq@XeAh@q@\\yBdAi@T??u@\\m@ZmBlAgCjA{@h@w@l@MJKH[Zw@|@wLzMSTQR[XoE~E]`@sDtDoEzEe@d@sAvAsD~DmPxQsF~F}AtA]ZOLOLm@`@IF??yAxAWXkCzCeAtA{HfJiIzIcLzLuC`D}F`G??{@z@uBxB{@`AkCfCgA`Ac@Xw@`@i@Vy@XiAZwDdAqA`@o@PiAj@gAr@aAt@wAnAUV}@fAKJKNEDMNmAvA}BjC_EbFyDnEgDzDgAfAeBnB_F|FIHIJw@~@aAfAcAhAcDrDSTSTg@n@a@b@iL`NmEhFiCxCqLjNcDpD}DvE}HtIgAz@oAz@IDgAh@_A`@_ARy@LkBHy@@wAIuAOqBa@YEsWgE{D}@yCw@}IoCeDcAgF_BkB_@_AK}BEyHXeE@{@@sEJkCF]Bg@@}AFM@a@D_AN??]FSDUBm@FqCd@}@VgAd@eAr@m@f@g@j@{@hAoAhBkBdD[h@}A|Bq@hAUZ}B|Cm@d@m@\\{@XY@MC_@QqBq@{F_C??g@ScnAl@"
     },
     %Site.ShuttleService.Shape{
       direction_id: 1,
       id: "933_0010-70082-70072-2",
       is_shuttle_route: false,
       polyline: "}{paGtyzpLOfD??ATIrB_Ap^s@jRWlJYdHOhGIpBKjAk@bBe@bAaDnGkBzD??m@pAwMpVgArCsBdGkEbPc@nAYb@a@d@eChDwG|HMPEFwCzNK^KVMJIB_@?iAO??kAOmB]gBGm@A[HSL[^a@|@e@fAo@r@iBh@aB^}@BkAGaNq@am@iDQ???sMUyQ~@uCJsBDo@Pc@Xm@l@m@bAaAvBu@pB]hAOx@UdAKr@??i@dD_@lDKlBOpCQjESvHKtEYbJg@rUGjCRnDFv@NbAPjAHRHLLFt@Nt@Vt@j@f@|@Vt@@DR~B@b@HpAFxN"
     }
   ],
   stops: [
     %Site.ShuttleService.Stop{
       id: "place-alfcl",
       latitude: 42.395428,
       longitude: -71.142483,
       name: "Alewife",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-davis",
       latitude: 42.39674,
       longitude: -71.121815,
       name: "Davis",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-portr",
       latitude: 42.3884,
       longitude: -71.119149,
       name: "Porter",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-harsq",
       latitude: 42.373362,
       longitude: -71.118956,
       name: "Harvard",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-cntsq",
       latitude: 42.365486,
       longitude: -71.103802,
       name: "Central",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-knncl",
       latitude: 42.362491,
       longitude: -71.086177,
       name: "Kendall/MIT",
       type: :rail_affected
     },
     %Site.ShuttleService.Stop{
       id: "place-chmnl",
       latitude: 42.361166,
       longitude: -71.070628,
       name: "Charles/MGH",
       type: :rail_affected
     },
     %Site.ShuttleService.Stop{
       id: "place-pktrm",
       latitude: 42.356395,
       longitude: -71.062424,
       name: "Park Street",
       type: :rail_affected
     },
     %Site.ShuttleService.Stop{
       id: "place-dwnxg",
       latitude: 42.355518,
       longitude: -71.060225,
       name: "Downtown Crossing",
       type: :rail_affected
     },
     %Site.ShuttleService.Stop{
       id: "place-sstat",
       latitude: 42.352271,
       longitude: -71.055242,
       name: "South Station",
       type: :rail_affected
     },
     %Site.ShuttleService.Stop{
       id: "place-brdwy",
       latitude: 42.342622,
       longitude: -71.056967,
       name: "Broadway",
       type: :rail_affected
     },
     %Site.ShuttleService.Stop{
       id: "place-andrw",
       latitude: 42.330154,
       longitude: -71.057655,
       name: "Andrew",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-jfk",
       latitude: 42.320685,
       longitude: -71.052391,
       name: "JFK/UMass",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-shmnl",
       latitude: 42.31129,
       longitude: -71.053331,
       name: "Savin Hill",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-fldcr",
       latitude: 42.300093,
       longitude: -71.061667,
       name: "Fields Corner",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-smmnl",
       latitude: 42.293126,
       longitude: -71.065738,
       name: "Shawmut",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-asmnl",
       latitude: 42.284652,
       longitude: -71.064489,
       name: "Ashmont",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-nqncy",
       latitude: 42.275275,
       longitude: -71.029583,
       name: "North Quincy",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-wlsta",
       latitude: 42.266514,
       longitude: -71.020337,
       name: "Wollaston",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-qnctr",
       latitude: 42.251809,
       longitude: -71.005409,
       name: "Quincy Center",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-qamnl",
       latitude: 42.233391,
       longitude: -71.007153,
       name: "Quincy Adams",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "place-brntn",
       latitude: 42.207854,
       longitude: -71.001138,
       name: "Braintree",
       type: :rail_unaffected
     },
     %Site.ShuttleService.Stop{
       id: "9070071",
       latitude: 42.362551,
       longitude: -71.086826,
       name: "Kendall/MIT",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "9070073",
       latitude: 42.361003,
       longitude: -71.068975,
       name: "Charles/MGH",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "9070076",
       latitude: 42.356164,
       longitude: -71.062487,
       name: "Park Street",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "6564",
       latitude: 42.352253,
       longitude: -71.054774,
       name: "Summer St @ South Station - Red Line entrance",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "150",
       latitude: 42.342652,
       longitude: -71.057065,
       name: "Broadway Station - Red Line",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "151",
       latitude: 42.342664,
       longitude: -71.056949,
       name: "W Broadway @ Broadway Station",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "892",
       latitude: 42.3526,
       longitude: -71.05501,
       name: "Summer St @ Atlantic Ave",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "6551",
       latitude: 42.355286,
       longitude: -71.056411,
       name: "Federal St @ Franklin St",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "4510",
       latitude: 42.359634,
       longitude: -71.059706,
       name: "Cambridge St @ Government Ctr Sta",
       type: :shuttle
     },
     %Site.ShuttleService.Stop{
       id: "9070074",
       latitude: 42.361222,
       longitude: -71.069966,
       name: "Charles/MGH",
       type: :shuttle
     }
   ]
 }
</pre>
</details>

### CODE REVIEW

**Comment by digitalcora** [comment]
@phildarnowsky Thanks! Cristen pointed out yesterday that the word "Service" is kind of overloaded; in particular we have a `Service` domain model that this doesn't correspond to at all. Any ideas on a better name? Maybe `ShuttleDiversion`?

File: apps/site/lib/site/shuttle_service.ex [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+  defmodule Shape do
+    @moduledoc "Derivative of an API `shape` tailored for drawing a shuttle service map."
+
+    @enforce_keys [:id, :direction_id, :polyline, :is_shuttle_route]
+    defstruct @enforce_keys
+
+    @type t :: %__MODULE__{
+            id: String.t(),
+            direction_id: integer,
```
[the remark this answers is not in the document]
Comment (digitalcora) [reply apps/site/lib/site/shuttle_service.ex]: Good catch!

File: apps/site/lib/site/shuttle_service.ex [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  defstruct @enforce_keys
+
+  @type t :: %__MODULE__{
+          shapes: [Shape.t()],
+          stops: [Stop.t()]
+        }
+
+  @doc "Indicates whether there are any shuttle services on the given routes at a given time."
+  @spec any?([Route.id_t()], DateTime.t()) :: boolean
+  def any?(route_ids, time \\ Util.now()), do: Enum.any?(ongoing_shuttle_alerts(route_ids, time))
```
[the remark this answers is not in the document]
Comment (digitalcora) [reply apps/site/lib/site/shuttle_service.ex]: `active?` might be better. Coming up with good names is hard 😓

**Comment by digitalcora** [comment]
@thecristen I addressed your feedback and decoded the polylines (in e9ea14fbce32f858e84d54c22a953a3decbc781d), if it looks good I will fixup and merge 🙂

File: apps/site/lib/site/shuttle_diversion.ex [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  defmodule Shape do
+    @moduledoc "Derivative of an API `shape` tailored for drawing a shuttle diversion map."
+
+    @enforce_keys [:id, :direction_id, :polyline, :is_shuttle_route]
+    defstruct @enforce_keys
+
+    @type t :: %__MODULE__{
+            id: String.t(),
+            direction_id: 0 | 1,
+            polyline: [[float]],
```
Comment (thecristen) [inline apps/site/lib/site/shuttle_diversion.ex]: Question - should this indicate a list of lat/lng positions? Or is this the encoded version?

Comment (digitalcora) [reply apps/site/lib/site/shuttle_diversion.ex]: This is the decoded polyline, which is a list of lists of floats. Technically the lists of floats always contain exactly 2 elements (lat and long), but there is no way to encode that in the typespec. The idea is that if you have a fixed-length list, what you actually want is a tuple, and that _can_ be encoded as `{float, float}`. However, tuples can't be JSON-encoded.

**Comment by thecristen** [comment]
@digitalcora Looks good to me!

**Pull request merged**